hivemind/compression/adaptive.py

```diff
@@ -0,0 +1,68 @@
+from abc import ABC
+from typing import Mapping, Sequence, Union
+
+import torch
+
+import hivemind
+from hivemind.compression.base import CompressionBase, CompressionInfo, Key, NoCompression, TensorRole
+from hivemind.proto import runtime_pb2
+
+
+class AdaptiveCompressionBase(CompressionBase, ABC):
+    def extract(self, serialized_tensor: runtime_pb2.Tensor) -> torch.Tensor:
+        return hivemind.compression.deserialize_torch_tensor(serialized_tensor)
```

Comment:
Let's add an abstract method `choose_strategy(self, info: CompressionInfo) -> Compression` (don't forget to mark it with `@abstractmethod`) and implement `estimate_compression_ratio()` and `compress()` in this class.

Then, let's provide an implementation only for `choose_strategy` in the subclasses.

This will help to get rid of the copy-paste below.